workflow/train_test_split.py

```diff
@@ -0,0 +1,167 @@
+import pytest
+
+
+def train_test_split(
+    df,
+    unique_key,
+    test_size=0.2,
+    seed=0,
+    stratify=None
+):
+    if stratify is None:
+        test_df = get_test_split(df, unique_key, test_size, seed)
+    else:
+        test_df = stratified_split(df, unique_key, stratify, test_size, seed)
+
+    train_df = df[~df[unique_key].isin(test_df[unique_key])]
+
+    return train_df, test_df
+
+
+def get_test_split(df, unique_key, test_size, seed=0, resolution=100000):
+    split_unique_key = df[unique_key].apply(
+        lambda x: ((hash(hash(x) + seed) % resolution))
+    )
+    return df[split_unique_key < resolution * test_size]
+
+
+def stratified_split(df, unique_key, stratify, test_size, seed):
+
+    test_dfs = []
+    for label in get_labels(df, stratify):
+        if type(stratify) is str:
+            df_part = df[df[stratify] == label]
+        else:
+            df_part = df[df[label].notnull()]
+        test_dfs.append(get_test_split(df_part, unique_key, test_size, seed))
+        df = df[~df[unique_key].isin(df_part[unique_key])]
+    assert len(df) == 0, f'Unable to stratify on {stratify}'
+
+    return pd.concat(test_dfs)
+
+
+def get_labels(df, stratify):
+
+    if type(stratify) is str:
+        return sorted(df[stratify].unique())
+    else:
+        if not isinstance(stratify, Iterable):
+            raise ValueError(
+                f'Argument stratify must be str or iterable, was {stratify}'
+            )
+        return sorted(stratify)
+
+
+@pytest.fixture
+def test_df(num_data=10000):
+    return (
+        pd.DataFrame(dict(
+            unique_key=np.arange(0, num_data),
+            label_1=np.random.choice(['a', None], num_data),
+            label_2=np.random.choice(['b', None], num_data),
+            label_3=np.random.choice(['c', None], num_data),
+            multiclass=np.random.choice([1, 2, 3], num_data),
+        ))
+        .assign(unique_key=lambda df: df['unique_key'].astype(str))
+        .assign(no_label=lambda df: (
+            df['label_1'].isnull() &
+            df['label_2'].isnull() &
+            df['label_3'].isnull()
+        ))
+    )
+
+
+def test_non_stratified_split(test_df):
+    try:
+        train, test = train_test_split(
+            test_df,
+            'unique_key',
+            test_size=0.2,
+            seed=1,
+        )
+    except:
+        raise Exception('Failed to split without stratification.')
+
+
+def test_multiclass_stratify_split(test_df):
+    train, test = train_test_split(
+        test_df,
+        'unique_key',
+        test_size=0.2,
+        seed=1,
+        stratify='multiclass',
+    )
+    for label in test_df['multiclass'].unique():
+        ratio = (
+            len(test[test['multiclass'] == label])
+            / len(test_df[test_df['multiclass'] == label])
+        )
+        if ratio < 0.1 or ratio > 0.3:
+            raise Exception(f'Expected ratio close to 0.2, got {ratio}')
+        assert ratio < 0.3 and ratio > 0.1
```

Comment:
```suggestion
        assert 0.1 < ratio < 0.3
```
(actually works as expected in python)

Reply:
https://github.com/codewell/ml-workflow/pull/83